### PR TITLE
[internal] ta: keypairgen ecc/rsa: SKS_CKA_LABEL

### PR DESCRIPTION
**Pull request by ldts**
During keypairgen, the label can be used to let crypto know that a
specific pre-allocated key present in the SE must be pulled and
associated to this key. For that the label must be set as

--label SE_XXXXXXXX

The SE key will correspond to the 32bit hexadecimal value defined
after the SE_ tag

Signed-off-by: Jorge Ramirez-Ortiz <jorge@foundries.io>

### CODE REVIEW

File: ta/secure_key_services/src/processing_ec.c [hunk 1]
```diff
@@ -1274,6 +1275,15 @@ uint32_t generate_ec_keys(struct sks_attribute_head *proc_params,
 
 	TEE_InitValueAttribute(&tee_key_attr[0], TEE_ATTR_ECC_CURVE,
 				tee_curve, 0);
+	tee_count++;
+
+	/* pass the label to the private exponent */
+	rv = get_attribute_ptr(*pub_head, SKS_CKA_LABEL, &b_ptr, &b_size);
+	if (rv == SKS_OK) {
```
Comment (ldts) [inline ta/secure_key_services/src/processing_ec.c:1282]: um, this must not be done for every label, only those prefixed with SE_. let me fix.

File: ta/secure_key_services/src/processing_rsa.c [hunk 2]
```diff
@@ -469,6 +469,16 @@ uint32_t generate_rsa_keys(struct sks_attribute_head *proc_params,
 		tee_count++;
 	}
 
+	/* pass the label to the private exponent */
+	rv = get_attribute_ptr(*pub_head, SKS_CKA_LABEL,
+			       &a_ptr, &a_size);
+	if (rv == SKS_OK) {
+		TEE_InitRefAttribute(&tee_attrs[tee_count],
```
Comment (ldts) [inline ta/secure_key_services/src/processing_rsa.c:476]: same here, this shouldnt be done for all labels...

Comment (ldts) [reply ta/secure_key_services/src/processing_rsa.c:477]: fixed

File: ta/secure_key_services/src/processing_ec.c [hunk 3]
```diff
@@ -1274,6 +1275,16 @@ uint32_t generate_ec_keys(struct sks_attribute_head *proc_params,
...
+	tee_count++;
+
+	/* pass the label to the private exponent */
+	rv = get_attribute_ptr(*pub_head, SKS_CKA_LABEL, &b_ptr, &b_size);
+
+	if (rv == SKS_OK && b_size == 11 && !memcmp(b_ptr, "SE_", 3)) {
+		TEE_InitRefAttribute(&tee_key_attr[tee_count],
+				     TEE_ATTR_ECC_PRIVATE_VALUE, b_ptr, b_size);
+		tee_count++;
+	}
```
Comment (ricardosalveti) [inline ta/secure_key_services/src/processing_ec.c:1295]: Can you protect this extra logic with a ifdef or similar that only gets applied if optee-os was built with se050 support?

File: ta/secure_key_services/src/processing_ec.c [hunk 4]
```diff
@@ -1239,13 +1239,14 @@ uint32_t generate_ec_keys(struct sks_attribute_head *proc_params,
 			  struct sks_attrs_head **priv_head)
 {
 	uint32_t rv = 0;
-	void *a_ptr = NULL;
-	uint32_t a_size = 0;
+	void *a_ptr = NULL, *b_ptr = NULL;
+	uint32_t a_size = 0, b_size = 0;
```
Comment (etienne-lms) [inline ta/secure_key_services/src/processing_ec.c]: In optee, we tend to use 1 line per local var. 
That not a strong opinion but helps.

File: ta/secure_key_services/src/processing_ec.c (hunk 3 above)
Comment (etienne-lms) [reply ta/secure_key_services/src/processing_ec.c:1295]: Prefer `if (IS_ENABLED(CFG_SE05_...))`. Requires `#include <config.h>`.

File: ta/secure_key_services/src/processing_rsa.c [hunk 5]
```diff
@@ -9,6 +9,7 @@
 #include <tee_api_defines.h>
 #include <tee_internal_api.h>
 #include <tee_internal_api_extensions.h>
+#include <string.h>
```
Comment (etienne-lms) [inline ta/secure_key_services/src/processing_rsa.c]: minor: move before tee_api_defines.h.

File: ta/secure_key_services/src/processing_ec.c (hunk 4 above)
Comment (ldts) [reply ta/secure_key_services/src/processing_ec.c]: sure.

File: ta/secure_key_services/src/processing_ec.c (hunk 3 above)
Comment (ldts) [reply ta/secure_key_services/src/processing_ec.c:1295]: ok

File: ta/secure_key_services/src/processing_rsa.c (hunk 5 above)
Comment (ldts) [reply ta/secure_key_services/src/processing_rsa.c]: ok

**Review by ricardosalveti — APPROVED** [review]
LGTM

**Pull request merged**